crawlfrontier/contrib/canonicalsolvers/basic.py

```diff
@@ -0,0 +1,44 @@
+# -*- coding: utf-8 -*-
+from crawlfrontier.core.components import CanonicalSolver
+
+
+class BasicCanonicalSolver(CanonicalSolver):
+    """
+    Implements a simple CanonicalSolver taking always first URL from redirect chain, if there were redirects.
+    It allows easily to avoid leaking of requests in CF (e.g. when request issued by
+    :attr:`get_next_requests() <crawlfrontier.core.manager.FrontierManager.get_next_requests>` never matched in
+    :attr:`page_crawled() <crawlfrontier.core.manager.FrontierManager.page_crawled>`) at the price of duplicating
+    records in CF for pages having more than one URL or complex redirects chains.
+    """
+    def frontier_start(self):
+        pass
+
+    def frontier_stop(self):
+        pass
+
+    def add_seeds(self, seeds):
+        pass
+
+    def page_crawled(self, response, links):
+        pass
+
+    def request_error(self, page, error):
+        pass
+
+    def get_canonical_url(self, obj):
+        canonical_url = obj.url
+        canonical_fingerprint = obj.meta['fingerprint']
+        canonical_domain_fingerprint = obj.meta['domain']['fingerprint'] if 'domain' in obj.meta else str()
+        if 'redirect_urls' in obj.meta:
+            redirect_urls = obj.meta['redirect_urls']
+            redirect_fingerprints = obj.meta['redirect_fingerprints']
+            redirect_domains = obj.meta['redirect_domains']
```

Comment:
Well, this decision is full of trade-offs. This data is needed in backend because fingerprint, url and domain is used for record identification purposes and batch generation. The fingerprint and domain related logic is managed and configured in CF middlewares, so I didn't want to overengineer this dealing with settings in CanonicalSolver, etc. At the end of the day we save some CPU cycles caching those values.